// @storybook/core/dist/core-events/index.js
// src/core-events/index.ts
var _ = /* @__PURE__ */ ((E) => (E.CHANNEL_WS_DISCONNECT = "channelWSDisconnect", E.CHANNEL_CREATED = "channelCreated", E.CONFIG_ERROR = "co\
nfigError", E.STORY_INDEX_INVALIDATED = "storyIndexInvalidated", E.STORY_SPECIFIED = "storySpecified", E.SET_CONFIG = "setConfig", E.SET_STORIES =
"setStories", E.SET_INDEX = "setIndex", E.SET_CURRENT_STORY = "setCurrentStory", E.CURRENT_STORY_WAS_SET = "currentStoryWasSet", E.FORCE_RE_RENDER =
"forceReRender", E.FORCE_REMOUNT = "forceRemount", E.PRELOAD_ENTRIES = "preloadStories", E.STORY_PREPARED = "storyPrepared", E.DOCS_PREPARED =
"docsPrepared", E.STORY_CHANGED = "storyChanged", E.STORY_UNCHANGED = "storyUnchanged", E.STORY_RENDERED = "storyRendered", E.STORY_FINISHED =
"storyFinished", E.STORY_MISSING = "storyMissing", E.STORY_ERRORED = "storyErrored", E.STORY_THREW_EXCEPTION = "storyThrewException", E.STORY_RENDER_PHASE_CHANGED =
"storyRenderPhaseChanged", E.PLAY_FUNCTION_THREW_EXCEPTION = "playFunctionThrewException", E.UNHANDLED_ERRORS_WHILE_PLAYING = "unhandledErro\
rsWhilePlaying", E.UPDATE_STORY_ARGS = "updateStoryArgs", E.STORY_ARGS_UPDATED = "storyArgsUpdated", E.RESET_STORY_ARGS = "resetStoryArgs", E.
SET_FILTER = "setFilter", E.SET_GLOBALS = "setGlobals", E.UPDATE_GLOBALS = "updateGlobals", E.GLOBALS_UPDATED = "globalsUpdated", E.REGISTER_SUBSCRIPTION =
"registerSubscription", E.PREVIEW_KEYDOWN = "previewKeydown", E.PREVIEW_BUILDER_PROGRESS = "preview_builder_progress", E.SELECT_STORY = "sel\
ectStory", E.STORIES_COLLAPSE_ALL = "storiesCollapseAll", E.STORIES_EXPAND_ALL = "storiesExpandAll", E.DOCS_RENDERED = "docsRendered", E.SHARED_STATE_CHANGED =
"sharedStateChanged", E.SHARED_STATE_SET = "sharedStateSet", E.NAVIGATE_URL = "navigateUrl", E.UPDATE_QUERY_PARAMS = "updateQueryParams", E.
REQUEST_WHATS_NEW_DATA = "requestWhatsNewData", E.RESULT_WHATS_NEW_DATA = "resultWhatsNewData", E.SET_WHATS_NEW_CACHE = "setWhatsNewCache", E.
TOGGLE_WHATS_NEW_NOTIFICATIONS = "toggleWhatsNewNotifications", E.TELEMETRY_ERROR = "telemetryError", E.FILE_COMPONENT_SEARCH_REQUEST = "fil\
eComponentSearchRequest", E.FILE_COMPONENT_SEARCH_RESPONSE = "fileComponentSearchResponse", E.SAVE_STORY_REQUEST = "saveStoryRequest", E.SAVE_STORY_RESPONSE =
"saveStoryResponse", E.ARGTYPES_INFO_REQUEST = "argtypesInfoRequest", E.ARGTYPES_INFO_RESPONSE = "argtypesInfoResponse", E.CREATE_NEW_STORYFILE_REQUEST =
"createNewStoryfileRequest", E.CREATE_NEW_STORYFILE_RESPONSE = "createNewStoryfileResponse", E.TESTING_MODULE_CRASH_REPORT = "testingModuleC\
rashReport", E.TESTING_MODULE_PROGRESS_REPORT = "testingModuleProgressReport", E.TESTING_MODULE_RUN_REQUEST = "testingModuleRunRequest", E.TESTING_MODULE_RUN_ALL_REQUEST =
"testingModuleRunAllRequest", E.TESTING_MODULE_CANCEL_TEST_RUN_REQUEST = "testingModuleCancelTestRunRequest", E.TESTING_MODULE_CANCEL_TEST_RUN_RESPONSE =
"testingModuleCancelTestRunResponse", E.TESTING_MODULE_WATCH_MODE_REQUEST = "testingModuleWatchModeRequest", E.TESTING_MODULE_CONFIG_CHANGE =
"testingModuleConfigChange", E))(_ || {}), R = _, {
  CHANNEL_WS_DISCONNECT: S,
  CHANNEL_CREATED: T,
  CONFIG_ERROR: N,
  CREATE_NEW_STORYFILE_REQUEST: O,
  CREATE_NEW_STORYFILE_RESPONSE: A,
  CURRENT_STORY_WAS_SET: r,
  DOCS_PREPARED: D,
  DOCS_RENDERED: I,
  FILE_COMPONENT_SEARCH_REQUEST: C,
  FILE_COMPONENT_SEARCH_RESPONSE: o,
  FORCE_RE_RENDER: L,
  FORCE_REMOUNT: U,
  GLOBALS_UPDATED: P,
  NAVIGATE_URL: G,
  PLAY_FUNCTION_THREW_EXCEPTION: a,
  UNHANDLED_ERRORS_WHILE_PLAYING: Y,
  PRELOAD_ENTRIES: t,
  PREVIEW_BUILDER_PROGRESS: e,
  PREVIEW_KEYDOWN: H,
  REGISTER_SUBSCRIPTION: d,
  RESET_STORY_ARGS: W,
  SELECT_STORY: l,
  SET_CONFIG: i,
  SET_CURRENT_STORY: M,
  SET_FILTER: p,
  SET_GLOBALS: u,
  SET_INDEX: F,
  SET_STORIES: g,
  SHARED_STATE_CHANGED: s,
  SHARED_STATE_SET: y,
  STORIES_COLLAPSE_ALL: c,
  STORIES_EXPAND_ALL: h,
  STORY_ARGS_UPDATED: f,
  STORY_CHANGED: Q,
  STORY_ERRORED: x,
  STORY_INDEX_INVALIDATED: m,
  STORY_MISSING: V,
  STORY_PREPARED: w,
  STORY_RENDER_PHASE_CHANGED: B,
  STORY_RENDERED: X,
  STORY_FINISHED: q,
  STORY_SPECIFIED: b,
  STORY_THREW_EXCEPTION: K,
  STORY_UNCHANGED: n,
  UPDATE_GLOBALS: j,
  UPDATE_QUERY_PARAMS: k,
  UPDATE_STORY_ARGS: z,
  REQUEST_WHATS_NEW_DATA: J,
  RESULT_WHATS_NEW_DATA: Z,
  SET_WHATS_NEW_CACHE: $,
  TOGGLE_WHATS_NEW_NOTIFICATIONS: v,
  TELEMETRY_ERROR: EE,
  SAVE_STORY_REQUEST: _E,
  SAVE_STORY_RESPONSE: RE,
  ARGTYPES_INFO_REQUEST: SE,
  ARGTYPES_INFO_RESPONSE: TE,
  TESTING_MODULE_CRASH_REPORT: NE,
  TESTING_MODULE_PROGRESS_REPORT: OE,
  TESTING_MODULE_RUN_REQUEST: AE,
  TESTING_MODULE_RUN_ALL_REQUEST: rE,
  TESTING_MODULE_CANCEL_TEST_RUN_REQUEST: DE,
  TESTING_MODULE_CANCEL_TEST_RUN_RESPONSE: IE,
  TESTING_MODULE_WATCH_MODE_REQUEST: CE,
  TESTING_MODULE_CONFIG_CHANGE: oE
} = _;
export {
  SE as ARGTYPES_INFO_REQUEST,
  TE as ARGTYPES_INFO_RESPONSE,
  T as CHANNEL_CREATED,
  S as CHANNEL_WS_DISCONNECT,
  N as CONFIG_ERROR,
  O as CREATE_NEW_STORYFILE_REQUEST,
  A as CREATE_NEW_STORYFILE_RESPONSE,
  r as CURRENT_STORY_WAS_SET,
  D as DOCS_PREPARED,
  I as DOCS_RENDERED,
  C as FILE_COMPONENT_SEARCH_REQUEST,
  o as FILE_COMPONENT_SEARCH_RESPONSE,
  U as FORCE_REMOUNT,
  L as FORCE_RE_RENDER,
  P as GLOBALS_UPDATED,
  G as NAVIGATE_URL,
  a as PLAY_FUNCTION_THREW_EXCEPTION,
  t as PRELOAD_ENTRIES,
  e as PREVIEW_BUILDER_PROGRESS,
  H as PREVIEW_KEYDOWN,
  d as REGISTER_SUBSCRIPTION,
  J as REQUEST_WHATS_NEW_DATA,
  W as RESET_STORY_ARGS,
  Z as RESULT_WHATS_NEW_DATA,
  _E as SAVE_STORY_REQUEST,
  RE as SAVE_STORY_RESPONSE,
  l as SELECT_STORY,
  i as SET_CONFIG,
  M as SET_CURRENT_STORY,
  p as SET_FILTER,
  u as SET_GLOBALS,
  F as SET_INDEX,
  g as SET_STORIES,
  $ as SET_WHATS_NEW_CACHE,
  s as SHARED_STATE_CHANGED,
  y as SHARED_STATE_SET,
  c as STORIES_COLLAPSE_ALL,
  h as STORIES_EXPAND_ALL,
  f as STORY_ARGS_UPDATED,
  Q as STORY_CHANGED,
  x as STORY_ERRORED,
  q as STORY_FINISHED,
  m as STORY_INDEX_INVALIDATED,
  V as STORY_MISSING,
  w as STORY_PREPARED,
  X as STORY_RENDERED,
  B as STORY_RENDER_PHASE_CHANGED,
  b as STORY_SPECIFIED,
  K as STORY_THREW_EXCEPTION,
  n as STORY_UNCHANGED,
  EE as TELEMETRY_ERROR,
  DE as TESTING_MODULE_CANCEL_TEST_RUN_REQUEST,
  IE as TESTING_MODULE_CANCEL_TEST_RUN_RESPONSE,
  oE as TESTING_MODULE_CONFIG_CHANGE,
  NE as TESTING_MODULE_CRASH_REPORT,
  OE as TESTING_MODULE_PROGRESS_REPORT,
  rE as TESTING_MODULE_RUN_ALL_REQUEST,
  AE as TESTING_MODULE_RUN_REQUEST,
  CE as TESTING_MODULE_WATCH_MODE_REQUEST,
  v as TOGGLE_WHATS_NEW_NOTIFICATIONS,
  Y as UNHANDLED_ERRORS_WHILE_PLAYING,
  j as UPDATE_GLOBALS,
  k as UPDATE_QUERY_PARAMS,
  z as UPDATE_STORY_ARGS,
  R as default
};
