// handsontable/commonjs/3rdparty/walkontable/src/polyfilled.js
"use strict";

require("core-js/modules/es.object.get-own-property-descriptor");

exports.__esModule = true;
exports.default = void 0;

var _index = _interopRequireWildcard(require("./index"));

exports.ViewportColumnsCalculator = _index.ViewportColumnsCalculator;
exports.ViewportRowsCalculator = _index.ViewportRowsCalculator;
exports.CellCoords = _index.CellCoords;
exports.CellRange = _index.CellRange;
exports.ColumnFilter = _index.ColumnFilter;
exports.RowFilter = _index.RowFilter;
exports.DebugOverlay = _index.DebugOverlay;
exports.LeftOverlay = _index.LeftOverlay;
exports.TopOverlay = _index.TopOverlay;
exports.TopLeftCornerOverlay = _index.TopLeftCornerOverlay;
exports.BottomOverlay = _index.BottomOverlay;
exports.BottomLeftCornerOverlay = _index.BottomLeftCornerOverlay;
exports.Border = _index.Border;
exports.Core = _index.Core;
exports.Event = _index.Event;
exports.Overlays = _index.Overlays;
exports.Scroll = _index.Scroll;
exports.Selection = _index.Selection;
exports.Settings = _index.Settings;
exports.Table = _index.Table;
exports.TableRenderer = _index.TableRenderer;
exports.Viewport = _index.Viewport;

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

var _default = _index.default;
exports.default = _default;
