// handsontable/helpers/a11y.js
"use strict";

exports.__esModule = true;
const A11Y_TABINDEX = val => ['tabindex', val];
exports.A11Y_TABINDEX = A11Y_TABINDEX;
const A11Y_TREEGRID = () => ['role', 'treegrid'];
exports.A11Y_TREEGRID = A11Y_TREEGRID;
const A11Y_PRESENTATION = () => ['role', 'presentation'];
exports.A11Y_PRESENTATION = A11Y_PRESENTATION;
const A11Y_GRIDCELL = () => ['role', 'gridcell'];
exports.A11Y_GRIDCELL = A11Y_GRIDCELL;
const A11Y_ROWHEADER = () => ['role', 'rowheader'];
exports.A11Y_ROWHEADER = A11Y_ROWHEADER;
const A11Y_ROWGROUP = () => ['role', 'rowgroup'];
exports.A11Y_ROWGROUP = A11Y_ROWGROUP;
const A11Y_COLUMNHEADER = () => ['role', 'columnheader'];
exports.A11Y_COLUMNHEADER = A11Y_COLUMNHEADER;
const A11Y_ROW = () => ['role', 'row'];
exports.A11Y_ROW = A11Y_ROW;
const A11Y_MENU = () => ['role', 'menu'];
exports.A11Y_MENU = A11Y_MENU;
const A11Y_MENU_ITEM = () => ['role', 'menuitem'];
exports.A11Y_MENU_ITEM = A11Y_MENU_ITEM;
const A11Y_MENU_ITEM_CHECKBOX = () => ['role', 'menuitemcheckbox'];
exports.A11Y_MENU_ITEM_CHECKBOX = A11Y_MENU_ITEM_CHECKBOX;
const A11Y_COMBOBOX = () => ['role', 'combobox'];
exports.A11Y_COMBOBOX = A11Y_COMBOBOX;
const A11Y_LISTBOX = () => ['role', 'listbox'];
exports.A11Y_LISTBOX = A11Y_LISTBOX;
const A11Y_OPTION = () => ['role', 'option'];
exports.A11Y_OPTION = A11Y_OPTION;
const A11Y_CHECKBOX = () => ['role', 'checkbox'];
exports.A11Y_CHECKBOX = A11Y_CHECKBOX;
const A11Y_SCOPE_COL = () => ['scope', 'col'];
exports.A11Y_SCOPE_COL = A11Y_SCOPE_COL;
const A11Y_SCOPE_ROW = () => ['scope', 'row'];
exports.A11Y_SCOPE_ROW = A11Y_SCOPE_ROW;
const A11Y_TEXT = () => ['type', 'text'];
exports.A11Y_TEXT = A11Y_TEXT;
const A11Y_LABEL = val => ['aria-label', val];
exports.A11Y_LABEL = A11Y_LABEL;
const A11Y_HIDDEN = () => ['aria-hidden', 'true'];
exports.A11Y_HIDDEN = A11Y_HIDDEN;
const A11Y_DISABLED = () => ['aria-disabled', 'true'];
exports.A11Y_DISABLED = A11Y_DISABLED;
const A11Y_MULTISELECTABLE = () => ['aria-multiselectable', 'true'];
exports.A11Y_MULTISELECTABLE = A11Y_MULTISELECTABLE;
const A11Y_HASPOPUP = val => ['aria-haspopup', val];
exports.A11Y_HASPOPUP = A11Y_HASPOPUP;
const A11Y_ROWCOUNT = val => ['aria-rowcount', val];
exports.A11Y_ROWCOUNT = A11Y_ROWCOUNT;
const A11Y_COLCOUNT = val => ['aria-colcount', val];
exports.A11Y_COLCOUNT = A11Y_COLCOUNT;
const A11Y_ROWINDEX = val => ['aria-rowindex', val];
exports.A11Y_ROWINDEX = A11Y_ROWINDEX;
const A11Y_COLINDEX = val => ['aria-colindex', val];
exports.A11Y_COLINDEX = A11Y_COLINDEX;
const A11Y_EXPANDED = val => ['aria-expanded', val];
exports.A11Y_EXPANDED = A11Y_EXPANDED;
const A11Y_SORT = val => ['aria-sort', val];
exports.A11Y_SORT = A11Y_SORT;
const A11Y_READONLY = () => ['aria-readonly', 'true'];
exports.A11Y_READONLY = A11Y_READONLY;
const A11Y_INVALID = () => ['aria-invalid', 'true'];
exports.A11Y_INVALID = A11Y_INVALID;
const A11Y_CHECKED = val => ['aria-checked', val];
exports.A11Y_CHECKED = A11Y_CHECKED;
const A11Y_SELECTED = () => ['aria-selected', 'true'];
exports.A11Y_SELECTED = A11Y_SELECTED;
const A11Y_AUTOCOMPLETE = () => ['aria-autocomplete', 'list'];
exports.A11Y_AUTOCOMPLETE = A11Y_AUTOCOMPLETE;
const A11Y_CONTROLS = val => ['aria-controls', val];
exports.A11Y_CONTROLS = A11Y_CONTROLS;
const A11Y_ACTIVEDESCENDANT = val => ['aria-activedescendant', val];
exports.A11Y_ACTIVEDESCENDANT = A11Y_ACTIVEDESCENDANT;
const A11Y_LIVE = val => ['aria-live', val];
exports.A11Y_LIVE = A11Y_LIVE;
const A11Y_RELEVANT = val => ['aria-relevant', val];
exports.A11Y_RELEVANT = A11Y_RELEVANT;
const A11Y_SETSIZE = val => ['aria-setsize', val];
exports.A11Y_SETSIZE = A11Y_SETSIZE;
const A11Y_POSINSET = val => ['aria-posinset', val];
exports.A11Y_POSINSET = A11Y_POSINSET;
